ramips: Ubiquiti ER-X-SFP: add all the bits to enable SFP port eth5

vDorst opened the pull request:
Adding all the bits to enable SFP port eth5.

SFP cage of this device is connected via a AT8033 phy to port 5 of the switch.
This phy act as a RGMII-to-SerDes converter.

Also a I2C clock gate needs to be enabled in order to access the SFP module via I2C bus.
SFP cage also has module detect pin which is connected to I2C gpio expander.

With this patch the kernel/PHYLINK now can detect, readout and use the SFP module/port.

NOTE: SFP cage / AT8033 PHY only support 1000base-X encoding! This means that some SGMII modules can work and only at forced 1000MBit/full-duplex!

Bootlog
```
3: System Boot system code via Flash.
## Booting image at bfd40000 ...
   Image Name:   MIPS OpenWrt Linux-5.4.34
   Image Type:   MIPS Linux Kernel Image (uncompressed)
   Data Size:    2323462 Bytes =  2.2 MB
   Load Address: 80001000
   Entry Point:  80001000
....................................   Verifying Checksum ... OK
OK
No initrd
## Transferring control to Linux (at address 80001000) ...
## Giving linux memsize in MB, 256

Starting kernel ...



OpenWrt kernel loader for MIPS based SoC
Copyright (C) 2011 Gabor Juhos <juhosg@openwrt.org>
Decompressing kernel... done!
Starting kernel at 80001000...

[    0.000000] Linux version 5.4.34 (rene@pc-rene) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r12712+22-556b8581a1)) #0 SMP Fri Apr 24 07:24:10 2020
[    0.000000] SoC Type: MediaTek MT7621 ver:1 eco:3
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001992f (MIPS 1004Kc)
[    0.000000] MIPS: machine is Ubiquiti EdgeRouter X SFP
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] VPE topology {2,2} total 4
[    0.000000] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000]   HighMem  empty
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x000000000fffffff]
[    0.000000] percpu: Embedded 14 pages/cpu s26672 r8192 d22480 u57344
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 64960
[    0.000000] Kernel command line: console=ttyS0,57600 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 32768 (order: 5, 131072 bytes, linear)
[    0.000000] Inode-cache hash table entries: 16384 (order: 4, 65536 bytes, linear)
[    0.000000] Writing ErrCtl register=0001d090
[    0.000000] Readback ErrCtl register=0001d090
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 250864K/262144K available (5566K kernel code, 198K rwdata, 1212K rodata, 1268K init, 229K bss, 11280K reserved, 0K cma-reserved, 0K highmem)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 25 jiffies.
[    0.000000] NR_IRQS: 256
[    0.000000] random: get_random_bytes called from start_kernel+0x340/0x554 with crng_init=0
[    0.000000] CPU Clock: 880MHz
[    0.000000] clocksource: GIC: mask: 0xffffffffffffffff max_cycles: 0xcaf478abb4, max_idle_ns: 440795247997 ns
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 4343773742 ns
[    0.000009] sched_clock: 32 bits at 440MHz, resolution 2ns, wraps every 4880645118ns
[    0.015487] Calibrating delay loop... 583.68 BogoMIPS (lpj=1167360)
[    0.055841] pid_max: default: 32768 minimum: 301
[    0.065181] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.079587] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.097671] rcu: Hierarchical SRCU implementation.
[    0.107773] smp: Bringing up secondary CPUs ...
[    6.640385] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    6.640395] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    6.640407] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    6.640507] CPU1 revision is: 0001992f (MIPS 1004Kc)
[    0.144960] Synchronize counters for CPU 1: done.
[    6.731434] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    6.731442] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    6.731450] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    6.731503] CPU2 revision is: 0001992f (MIPS 1004Kc)
[    0.239397] Synchronize counters for CPU 2: done.
[    6.822544] Primary instruction cache 32kB, VIPT, 4-way, linesize 32 bytes.
[    6.822551] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    6.822559] MIPS secondary cache 256kB, 8-way, linesize 32 bytes.
[    6.822616] CPU3 revision is: 0001992f (MIPS 1004Kc)
[    0.326993] Synchronize counters for CPU 3: done.
[    0.386598] smp: Brought up 1 node, 4 CPUs
[    0.399114] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.418396] futex hash table entries: 1024 (order: 3, 32768 bytes, linear)
[    0.432324] pinctrl core: initialized pinctrl subsystem
[    0.444143] NET: Registered protocol family 16
[    0.480773] random: fast init done
[    0.487541] clocksource: Switched to clocksource GIC
[    0.499369] NET: Registered protocol family 2
[    0.508587] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 6144 bytes, linear)
[    0.525103] TCP established hash table entries: 2048 (order: 1, 8192 bytes, linear)
[    0.540303] TCP bind hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.554474] TCP: Hash tables configured (established 2048 bind 2048)
[    0.567198] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.580083] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.594163] NET: Registered protocol family 1
[    0.602719] PCI: CLS 0 bytes, default 32
[    0.699502] 4 CPUs re-calibrate udelay(lpj = 1167360)
[    0.711185] workingset: timestamp_bits=14 max_order=16 bucket_order=2
[    0.735466] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.746955] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.768798] GPIO line 487 (sfp_i2c_clk_gate) hogged as output/high
[    0.781289] mt7621_gpio 1e000600.gpio: registering 32 gpios
[    0.792553] mt7621_gpio 1e000600.gpio: registering 32 gpios
[    0.803826] mt7621_gpio 1e000600.gpio: registering 32 gpios
[    0.815618] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    0.829381] printk: console [ttyS0] disabled
[    0.837823] 1e000c00.uartlite: ttyS0 at MMIO 0x1e000c00 (irq = 19, base_baud = 3125000) is a 16550A
[    0.855760] printk: console [ttyS0] enabled
[    0.855760] printk: console [ttyS0] enabled
[    0.872306] printk: bootconsole [early0] disabled
[    0.872306] printk: bootconsole [early0] disabled
[    0.893197] mt7621-nand 1e003000.nand: Using programmed access timing: 31c07388
[    0.908024] nand: device found, Manufacturer ID: 0x01, Chip ID: 0xda
[    0.920674] nand: AMD/Spansion S34ML02G2
[    0.928482] nand: 256 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 128
[    0.943733] mt7621-nand 1e003000.nand: ECC strength adjusted to 12 bits
[    0.956938] mt7621-nand 1e003000.nand: Using programmed access timing: 21005134
[    0.971513] mt7621-nand 1e003000.nand: Using programmed access timing: 21005134
[    0.986066] Scanning device for bad blocks
[    4.950878] mt7621-nand 1e003000.nand: Uncorrectable ECC error at page 130944.0
[    4.965599] mt7621-nand 1e003000.nand: Uncorrectable ECC error at page 130944.1
[    4.980316] mt7621-nand 1e003000.nand: Uncorrectable ECC error at page 130944.2
[    4.995031] mt7621-nand 1e003000.nand: Uncorrectable ECC error at page 130944.3
[    5.011078] mt7621-nand 1e003000.nand: Uncorrectable ECC error at page 131008.0
[    5.025796] mt7621-nand 1e003000.nand: Uncorrectable ECC error at page 131008.1
[    5.040514] mt7621-nand 1e003000.nand: Uncorrectable ECC error at page 131008.2
[    5.055234] mt7621-nand 1e003000.nand: Uncorrectable ECC error at page 131008.3
[    5.071132] 6 fixed-partitions partitions found on MTD device mt7621-nand
[    5.084651] Creating 6 MTD partitions on "mt7621-nand":
[    5.095061] 0x000000000000-0x000000080000 : "u-boot"
[    5.106202] 0x000000080000-0x0000000e0000 : "u-boot-env"
[    5.118008] 0x0000000e0000-0x000000140000 : "factory"
[    5.129321] 0x000000140000-0x000000440000 : "kernel1"
[    5.140677] 0x000000440000-0x000000740000 : "kernel2"
[    5.151947] 0x000000740000-0x00000ff00000 : "ubi"
[    5.165066] libphy: Fixed MDIO Bus: probed
[    5.198018] libphy: mdio: probed
[    5.204660] mt7530 mdio-bus:1f: MT7530 adapts as multi-chip module
[    5.237142] mtk_soc_eth 1e100000.ethernet dsa: mediatek frame engine at 0xbe100000, irq 20
[    5.254929] i2c-mt7621 1e000900.i2c: clock 100 kHz
[    5.268009] NET: Registered protocol family 10
[    5.278423] Segment Routing with IPv6
[    5.285845] NET: Registered protocol family 17
[    5.295004] 8021q: 802.1Q VLAN Support v1.8
[    5.305128] mt7530 mdio-bus:1f: MT7530 adapts as multi-chip module
[    6.291883] libphy: dsa slave smi: probed
[    6.311600] mt7530 mdio-bus:1f eth0 (uninitialized): PHY [dsa-0.0:00] driver [Generic PHY]
[    6.343589] mt7530 mdio-bus:1f eth1 (uninitialized): PHY [dsa-0.0:01] driver [Generic PHY]
[    6.375593] mt7530 mdio-bus:1f eth2 (uninitialized): PHY [dsa-0.0:02] driver [Generic PHY]
[    6.407589] mt7530 mdio-bus:1f eth3 (uninitialized): PHY [dsa-0.0:03] driver [Generic PHY]
[    6.439594] mt7530 mdio-bus:1f eth4 (uninitialized): PHY [dsa-0.0:04] driver [Generic PHY]
[    6.483581] mt7530 mdio-bus:1f eth5 (uninitialized): PHY [mdio-bus:07] driver [Atheros 8031 ethernet]
[    6.523538] mt7530 mdio-bus:1f: configuring for fixed/rgmii link mode
[    7.147532] mt7530 mdio-bus:1f: Link is Up - 1Gbps/Full - flow control off
[    7.231565] DSA: tree 0 setup
[    7.238652] UBI: auto-attach mtd5
[    7.245282] ubi0: attaching mtd5
[    9.813041] ubi0: scanning is finished
[    9.839418] ubi0: attached mtd5 (name "ubi", size 247 MiB)
[    9.850373] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    9.864063] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    9.877575] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[    9.891433] ubi0: good PEBs: 1982, bad PEBs: 0, corrupted PEBs: 0
[    9.903568] ubi0: user volume: 2, internal volumes: 1, max. volumes count: 128
[    9.917954] ubi0: max/mean erase counter: 18/12, WL threshold: 4096, image sequence number: 1553339802
[    9.936486] ubi0: available PEBs: 0, total reserved PEBs: 1982, PEBs reserved for bad PEB handling: 40
[    9.955043] ubi0: background thread "ubi_bgt0d" started, PID 458
[    9.956917] block ubiblock0_0: created from ubi0:0(rootfs)
[    9.977940] ubiblock: device ubiblock0_0 (rootfs) set to be root filesystem
[    9.991818] hctosys: unable to open rtc device (rtc0)
[   10.007426] VFS: Mounted root (squashfs filesystem) readonly on device 254:0.
[   10.025973] Freeing unused kernel memory: 1268K
[   10.035016] This architecture does not have kernel memory protection.
[   10.047840] Run /sbin/init as init process
[   10.506580] init: Console is alive
[   10.513612] init: - watchdog -
[   11.184023] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[   11.286281] usbcore: registered new interface driver usbfs
[   11.297337] usbcore: registered new interface driver hub
[   11.308103] usbcore: registered new device driver usb
[   11.322116] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[   11.347881] init: - preinit -
[   11.965744] mtk_soc_eth 1e100000.ethernet dsa: configuring for fixed/rgmii link mode
[   11.981679] mtk_soc_eth 1e100000.ethernet dsa: Link is Up - 1Gbps/Full - flow control rx/tx
[   11.998375] IPv6: ADDRCONF(NETDEV_CHANGE): dsa: link becomes ready
[   12.207029] random: jshn: uninitialized urandom read (4 bytes read)
[   12.321160] random: jshn: uninitialized urandom read (4 bytes read)
[   12.356477] random: jshn: uninitialized urandom read (4 bytes read)
[   12.895549] mt7530 mdio-bus:1f eth0: configuring for phy/gmii link mode
[   12.935667] 8021q: adding VLAN 0 to HW filter on device eth0
Press the [f] key and hit [enter] to enter failsafe mode
Press the [1], [2], [3] or [4] key and hit [enter] to select the debug level
[   17.251557] mt7530 mdio-bus:1f eth0: Link is Up - 1Gbps/Full - flow control rx/tx
[   17.266511] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   17.476482] UBIFS (ubi0:1): Mounting in unauthenticated mode
[   17.488106] UBIFS (ubi0:1): background thread "ubifs_bgt0_1" started, PID 564
[   17.530529] urandom_read: 4 callbacks suppressed
[   17.530541] random: procd: uninitialized urandom read (4 bytes read)
[   17.695840] UBIFS (ubi0:1): UBIFS: mounted UBI device 0, volume 1, name "rootfs_data"
[   17.711454] UBIFS (ubi0:1): LEB size: 126976 bytes (124 KiB), min./max. I/O unit sizes: 2048 bytes/2048 bytes
[   17.731205] UBIFS (ubi0:1): FS size: 241254400 bytes (230 MiB, 1900 LEBs), journal size 12062720 bytes (11 MiB, 95 LEBs)
[   17.752854] UBIFS (ubi0:1): reserved for root: 4952683 bytes (4836 KiB)
[   17.766042] UBIFS (ubi0:1): media format: w4/r0 (latest is w5/r0), UUID E1C16F79-A564-435F-888D-04FB8F03A8B9, small LPT model
[   17.790750] mount_root: overlay filesystem has not been fully initialized yet
[   17.805571] mount_root: switching to ubifs overlay
[   17.831035] overlayfs: upper fs does not support xattr, falling back to index=off and metacopy=off.
- config restore -
[   17.876960] urandom-seed: Seed file not found (/etc/urandom.seed)
[   17.995548] mt7530 mdio-bus:1f eth0: Link is Down
[   18.227599] procd: - early -
[   18.233446] procd: - watchdog -
[   18.771699] procd: - watchdog -
[   18.778304] procd: - ubus -
[   18.828887] random: ubusd: uninitialized urandom read (4 bytes read)
[   18.843599] random: ubusd: uninitialized urandom read (4 bytes read)
[   18.857574] procd: - init -
Please press Enter to activate this console.
[   19.282401] kmodloader: loading kernel modules from /etc/modules.d/*
[   19.368359] i2c /dev entries driver
[   19.378175] pca953x 0-0025: 0-0025 supply vcc not found, using dummy regulator
[   19.463933] pca953x 0-0025: interrupt support not compiled in
[   19.481101] Loading modules backported from Linux version v5.4.27-0-g585e0cc08069
[   19.496040] Backport generated by backports.git v5.4.27-1-0-gf6e8852f
[   19.520934] libphy: SFP I2C Bus: probed
[   19.538888] urngd: v1.0.2 started.
[   19.543625] sfp sfp_eth5: Host maximum power 1.0W
[   19.579551] sfp sfp_eth5: No tx_disable pin: SFP modules will always be emitting.
[   19.600394] wireguard: WireGuard 1.0.20200401 loaded. See www.wireguard.com for information.
[   19.617227] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[   19.643908] xt_time: kernel timezone is -0000
[   19.683436] random: crng init done
[   19.738315] PPP generic driver version 2.4.2
[   19.748561] NET: Registered protocol family 24
[   19.770034] kmodloader: done loading kernel modules from /etc/modules.d/*
[   20.115548] sfp sfp_eth5: module FS               SFP-GE-BX        rev      sn <snip>      dc 190803
[   20.134970] sfp sfp_eth5: Host does not support 2.0W modules, module left in power mode 1
[   20.151271] Atheros 8031 ethernet mdio-bus:07: SFP interface 1000base-x
[   25.937380] mtk_soc_eth 1e100000.ethernet dsa: Link is Down
[   25.957057] mtk_soc_eth 1e100000.ethernet dsa: configuring for fixed/rgmii link mode
[   25.972927] mtk_soc_eth 1e100000.ethernet dsa: Link is Up - 1Gbps/Full - flow control rx/tx
[   26.047563] mt7530 mdio-bus:1f eth0: configuring for phy/gmii link mode
[   26.091728] 8021q: adding VLAN 0 to HW filter on device eth0
[   26.411738] IPv6: ADDRCONF(NETDEV_CHANGE): dsa: link becomes ready
[   26.424673] br-lan: port 1(eth0) entered blocking state
[   26.435100] br-lan: port 1(eth0) entered disabled state
[   26.467564] device eth0 entered promiscuous mode
[   26.476771] device dsa entered promiscuous mode
[   26.923624] mt7530 mdio-bus:1f eth1: configuring for phy/gmii link mode
[   26.963984] 8021q: adding VLAN 0 to HW filter on device eth1
[   27.300037] br-lan: port 2(eth1) entered blocking state
[   27.310483] br-lan: port 2(eth1) entered disabled state
[   27.367552] device eth1 entered promiscuous mode
[   27.807543] mt7530 mdio-bus:1f eth2: configuring for phy/gmii link mode
[   27.847739] 8021q: adding VLAN 0 to HW filter on device eth2
[   28.184126] br-lan: port 3(eth2) entered blocking state
[   28.194603] br-lan: port 3(eth2) entered disabled state
[   28.275556] device eth2 entered promiscuous mode
[   28.723550] mt7530 mdio-bus:1f eth3: configuring for phy/gmii link mode
[   28.763716] 8021q: adding VLAN 0 to HW filter on device eth3
[   29.100033] br-lan: port 4(eth3) entered blocking state
[   29.110477] br-lan: port 4(eth3) entered disabled state
[   29.215578] device eth3 entered promiscuous mode
[   29.515554] mt7530 mdio-bus:1f eth0: Link is Up - 1Gbps/Full - flow control rx/tx
[   29.647635] br-lan: port 1(eth0) entered blocking state
[   29.658055] br-lan: port 1(eth0) entered forwarding state
[   29.715602] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   29.799550] mt7530 mdio-bus:1f eth4: configuring for phy/gmii link mode
[   29.839728] 8021q: adding VLAN 0 to HW filter on device eth4
[   30.184091] br-lan: port 5(eth4) entered blocking state
[   30.194521] br-lan: port 5(eth4) entered disabled state
[   30.331560] device eth4 entered promiscuous mode
[   30.783553] mt7530 mdio-bus:1f eth5: configuring for phy/rgmii-rxid link mode
[   30.823737] 8021q: adding VLAN 0 to HW filter on device eth5
[   30.859555] at803x_config_aneg: fiber
[   31.119564] mt7530 mdio-bus:1f eth5: Link is Up - 1Gbps/Full - flow control rx/tx
[   31.236188] IPv6: ADDRCONF(NETDEV_CHANGE): eth5: link becomes ready
[   33.311576] mt7530 mdio-bus:1f eth3: Link is Up - 1Gbps/Full - flow control rx/tx
[   33.326563] br-lan: port 4(eth3) entered blocking state
[   33.336995] br-lan: port 4(eth3) entered forwarding state
```

ifconfig output
```
root@OpenWrt:/# ifconfig
br-lan    Link encap:Ethernet  HWaddr 78:8A:20:07:52:00
          inet addr:192.168.1.1  Bcast:192.168.1.255  Mask:255.255.255.0
          inet6 addr: fdd3:5e2f:da83::1/60 Scope:Global
          inet6 addr: <snip>:500::1/60 Scope:Global
          inet6 addr: fe80::7a8a:20ff:fe07:5200/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:40 errors:0 dropped:0 overruns:0 frame:0
          TX packets:17 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:7167 (6.9 KiB)  TX bytes:2710 (2.6 KiB)

dsa       Link encap:Ethernet  HWaddr 78:8A:20:07:52:00
          inet6 addr: fe80::7a8a:20ff:fe07:5200/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:140 errors:0 dropped:0 overruns:0 frame:0
          TX packets:71 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:21410 (20.9 KiB)  TX bytes:9267 (9.0 KiB)
          Interrupt:20

eth0      Link encap:Ethernet  HWaddr 78:8A:20:07:52:00
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:40 errors:0 dropped:0 overruns:0 frame:0
          TX packets:19 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:7167 (6.9 KiB)  TX bytes:2890 (2.8 KiB)

eth1      Link encap:Ethernet  HWaddr 78:8A:20:07:52:01
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

eth2      Link encap:Ethernet  HWaddr 78:8A:20:07:52:02
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

eth3      Link encap:Ethernet  HWaddr 78:8A:20:07:52:03
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

eth4      Link encap:Ethernet  HWaddr 78:8A:20:07:52:04
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

eth5      Link encap:Ethernet  HWaddr 78:8A:20:07:52:05
          inet addr:192.168.2.15  Bcast:192.168.2.255  Mask:255.255.255.0
          inet6 addr: fe80::7a8a:20ff:fe07:5205/64 Scope:Link
          inet6 addr: <snip>:1:7a8a:20ff:fe07:5205/128 Scope:Global
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:100 errors:0 dropped:57 overruns:0 frame:0
          TX packets:36 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:11723 (11.4 KiB)  TX bytes:4133 (4.0 KiB)

lo        Link encap:Local Loopback
          inet addr:127.0.0.1  Mask:255.0.0.0
          inet6 addr: ::1/128 Scope:Host
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:20 errors:0 dropped:0 overruns:0 frame:0
          TX packets:20 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:1893 (1.8 KiB)  TX bytes:1893 (1.8 KiB)
```
   
ethtool SFP module Query/Decode Module EEPROM information and optical diagnostics if available       
```          
root@OpenWrt:/# ethtool -m eth5
        Identifier                                : 0x03 (SFP)
        Extended identifier                       : 0x04 (GBIC/SFP defined by 2-wire interface ID)
        Connector                                 : 0x07 (LC)
        Transceiver codes                         : 0x00 0x00 0x00 0x40 0x00 0x00 0x00 0x00 0x00
        Transceiver type                          : Ethernet: BASE-BX10
        Encoding                                  : 0x01 (8B/10B)
        BR, Nominal                               : 1300MBd
        Rate identifier                           : 0x00 (unspecified)
        Length (SMF,km)                           : 10km
        Length (SMF)                              : 10000m
        Length (50um)                             : 0m
        Length (62.5um)                           : 0m
        Length (Copper)                           : 0m
        Length (OM3)                              : 0m
        Laser wavelength                          : 1550nm
        Vendor name                               : FS
        Vendor OUI                                : 00:00:00
        Vendor PN                                 : SFP-GE-BX
        Vendor rev                                :
        Option values                             : 0x20 0x0a
        Option                                    : RX_LOS implemented
        Option                                    : TX_FAULT implemented
        Option                                    : Power level 3 requirement
        BR margin, max                            : 0%
        BR margin, min                            : 0%
        Vendor SN                                 : <snip>
        Date code                                 : 190803
        Optical diagnostics support               : Yes
        Laser bias current                        : 37.312 mA
        Laser output power                        : 0.2490 mW / -6.04 dBm
        Receiver signal average optical power     : 0.2058 mW / -6.87 dBm
        Module temperature                        : 54.83 degrees C / 130.70 degrees F
        Module voltage                            : 3.2059 V
        Alarm/warning flags implemented           : Yes
        Laser bias current high alarm             : Off
        Laser bias current low alarm              : Off
        Laser bias current high warning           : Off
        Laser bias current low warning            : Off
        Laser output power high alarm             : Off
        Laser output power low alarm              : Off
        Laser output power high warning           : Off
        Laser output power low warning            : Off
        Module temperature high alarm             : Off
        Module temperature low alarm              : Off
        Module temperature high warning           : Off
        Module temperature low warning            : Off
        Module voltage high alarm                 : Off
        Module voltage low alarm                  : Off
        Module voltage high warning               : Off
        Module voltage low warning                : Off
        Laser rx power high alarm                 : Off
        Laser rx power low alarm                  : Off
        Laser rx power high warning               : Off
        Laser rx power low warning                : Off
        Laser bias current high alarm threshold   : 100.000 mA
        Laser bias current low alarm threshold    : 0.000 mA
        Laser bias current high warning threshold : 95.000 mA
        Laser bias current low warning threshold  : 0.000 mA
        Laser output power high alarm threshold   : 0.7079 mW / -1.50 dBm
        Laser output power low alarm threshold    : 0.0891 mW / -10.50 dBm
        Laser output power high warning threshold : 0.6310 mW / -2.00 dBm
        Laser output power low warning threshold  : 0.1000 mW / -10.00 dBm
        Module temperature high alarm threshold   : 90.00 degrees C / 194.00 degrees F
        Module temperature low alarm threshold    : -45.00 degrees C / -49.00 degrees F
        Module temperature high warning threshold : 85.00 degrees C / 185.00 degrees F
        Module temperature low warning threshold  : -40.00 degrees C / -40.00 degrees F
        Module voltage high alarm threshold       : 3.7950 V
        Module voltage low alarm threshold        : 2.8050 V
        Module voltage high warning threshold     : 3.4650 V
        Module voltage low warning threshold      : 3.1350 V
        Laser rx power high alarm threshold       : 0.7079 mW / -1.50 dBm
        Laser rx power low alarm threshold        : 0.0035 mW / -24.56 dBm
        Laser rx power high warning threshold     : 0.6310 mW / -2.00 dBm
        Laser rx power low warning threshold      : 0.0040 mW / -23.98 dBm
```

Commit:
ramips: phy: at803x: support RGMII-to-SerDes and SFP support

Signed-off-by: DENG Qingfang <dengqf6@mail2.sysu.edu.cn>
Signed-off-by: René van Dorst <opensource@vdorst.com>

Commit:
kernel: backport support for clause 37 1000Base-X auto-negotiation

This patch is needed for clause 37 1000Base-X encoding used in many SFP modules.

Signed-off-by: DENG Qingfang <dengqf6@mail2.sysu.edu.cn>
Signed-off-by: René van Dorst <opensource@vdorst.com>

vDorst committed:
ramips: enable SFP port for Ubiquiti ER-X-SFP

SFP cage of this device is connected via a AT8031 phy to port 5 of the switch.
This phy act as a RGMII-to-SerDes converter.

Also a I2C clock gate needs to be enabled in order to access the SFP module via I2C bus.
SFP cage also has module detect pin which is connected to I2C gpio expander.

With this patch the kernel/PHYLINK now can detect, readout and use the SFP module/port.

NOTE: SFP cage / AT8033 PHY only support 1000base-X encoding!
This means that some SGMII modules can work and only at forced 1GBit/full-duplex!

Signed-off-by: René van Dorst <opensource@vdorst.com>

adrianschmutzler commented:
Hi, your first two commits lack commit description and Signed-off.

Despite, consider using "ramips:" prefix for your first patch, as this one will be in target directory.

adrianschmutzler commented:
Ah, I just realized that the first two patches are from @dengqf6 . So, obviously his Signed-off should be first, then.

dengqf6 commented on target/linux/ramips/mt7621/config-5.4:
```diff
@@ -281,6 +282,7 @@ CONFIG_SERIAL_8250_NR_UARTS=3
 CONFIG_SERIAL_8250_RUNTIME_UARTS=3
 CONFIG_SERIAL_MCTRL_GPIO=y
 CONFIG_SERIAL_OF_PLATFORM=y
+CONFIG_SFP=y
```
`SFP` and `MDIO_I2C` can be built as modules

vDorst replied:
> 
> 
> Hi, your first two commits lack commit description and Signed-off.
> 

I shall fix that.

> Despite, consider using "ramips:" prefix for your first patch, as this one will be in target directory.

OK.

vDorst replied:
How do you add a module to the image?
Only mark it as a module seems not enough to get is in the image.
I see that the module is build but not added to the image.

vDorst replied:
at803x needs to be build-in the kernel other wise it is loaded to late.
mt7530 is already loaded and initialized before at803x is loaded.

Maybe some other tricks to fix that?

dengqf6 replied:
at803x needs to be built-in while others are built-in, yes.

But SFP and MDIO_I2C (which SFP depends on) can be built as modules.

Further more, we can build every network related part as a module, this can decrease kernel size even more, and it may fit in HC5962's 2M kernel partition.

dengqf6 commented on target/linux/ramips/patches-5.4/991-at803x.patch:
```diff
@@ -0,0 +1,141 @@
+From 60ae82b0ea56c279be384b99cd2a42ae5ba7c5c7 Mon Sep 17 00:00:00 2001
```
Commit title's prefix should be "ramips:" if this is a platform specific patch.

vDorst replied:
Ok, shall change that.

dengqf6 commented:
Have you taken a look at `/etc/board.d/03_gpio_switches`? The GPIO numbers have changed

vDorst commented:
> Hi, your first two commits lack commit description and Signed-off.
> 
>Ah, I just realized that the first two patches are from @dengqf6 . So, obviously his Signed-off should be first, then. 

OK. Going to fix that.
@dengqf6 Are you OK if I put your Signed-off under your two patches/commits?

> Despite, consider using "ramips:" prefix for your first patch, as this one will be in target directory.
OK

dengqf6 commented:
I'm okay with that

vDorst committed:
kernel: add module to support SFP cages

Enables kernel SFP case support.

Signed-off-by: René van Dorst <opensource@vdorst.com>

most23 commented:
Regarding GPIO numbering (and probably sligthly of topic): on Edgerouter ER-X using 480 for "PoE Passthrough" works for me.

pmelange commented:
One thing which I think is a possible source for confusion is setting eth5 as the WAN port, where in the OEM firmware eth0 is the WAN port.  Wouldn't it make sense to keep it the same?

vDorst commented:
@most23 @dengqf6  I created a separate PR for fixing the POE gpio's. See https://github.com/openwrt/openwrt/pull/2910
I remove GPIO patch from this PR.

@pmelange That make sense. I'll change it back and add eth5 with the rest.

@dengqf6 I'll add SFP as a package and set AT803x as buildin for now.

ptpt52 commented on package/kernel/linux/modules/other.mk:
```diff
@@ -64,6 +64,26 @@ endef
+	CONFIG_SFP=m \
+	CONFIG_MDIO_I2C=m
+  FILES:= \
+	$(LINUX_DIR)/drivers/net/phy/at803x.ko \
+	$(LINUX_DIR)/drivers/net/phy/sfp.ko \
+	$(LINUX_DIR)/drivers/net/phy/mdio-i2c.ko
```
I think it is reasonable to separate these kmods into different packages, then use the dependencies to make it clear

vDorst replied:
I added a SFP package and set at803x as a build-in for now.

vDorst commented:
Changed:
* Removed the POE GPIO fix, is now a separate PR #2910.
* Restored eth0 as wan port, added eth5 as lan port.
* Made a SFP package to support SFP cage.

pmelange commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -42,10 +42,12 @@ ramips_setup_interfaces()
+		ucidef_set_interfaces_lan_wan "eth1 eth2 eth3 eth4 eth5" "eth0"
+		;;
 	*)
 		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" "wan"
```
Should maybe this line be changed to:
"lan1 lan2 lan3 lan4 SFP" "wan"
or
"lan1 lan2 lan3 lan4 lan5" "wan"

pmelange replied:
Or maybe "eth1 eth2 eth3 eth4 eth5/SFP" "eth0" like is printed on the device

pmelange replied:
but then seperately for the erx and erx-sfp

adrianschmutzler commented:
Hi, just some formal remarks:
- Please remove the dot at the end of the commit title(s)
- "add" after prefix should be lower-case
- The title of your second commit is too long. Maybe just "ramips: enable SFP port for Ubiquiti ER-X-SFP"?

vDorst commented:
Changes:
* Rebased
* All changes requested/suggested by @adrianschmutzler 
* Update bootlog in start post.

adrianschmutzler commented on target/linux/ramips/dts/mt7621_ubnt_edgerouter-x-sfp.dts:
```diff
@@ -5,13 +5,71 @@
+};
+
+&mdio {
+	ephy5: ethernet-phy@7 {
```
That should be labelled according to the reg property, i.e. phy7 (or ephy7).

vDorst replied:
Fixed.

adrianschmutzler commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -42,10 +42,12 @@ ramips_setup_interfaces()
 		ucidef_set_interface_lan "eth0 eth1 eth2 eth3 eth4"
 		;;
+	ubnt,edgerouter-x-sfp)
+		ucidef_set_interfaces_lan_wan "eth1 eth2 eth3 eth4 eth5" "eth0"
```
Where does the WAN port come from? There is nothing about it in the commit message.
IMO, that's a separate change and should be handled separately. (And I'd keep it all-LAN anyway).

pmelange replied:
In the OEM Firmware and in 19.x and previous versions of the firmware, eth0 was always WAN.  Perhaps it should be in the commit message, like you stated.  But from the user perspective, I think it makes sense to keep eth0 as WAN.

adrianschmutzler replied:
Well, from what I remember from my ER-X and from what's written in the manual, the default for eth0 is LAN with a static IP. If you then set up the device with the assistant, then there the default would be the eth0. So, to me, this sounds like all-LAN is default. That's also the reason why we changed this in https://github.com/openwrt/openwrt/commit/5acd1ed0be0d78847cd7d9d5599526f59babaf4d

So, from my perspective, the previous behavior in 19.xx has been "wrong" and the config break due to DSA is a perfect chance to make it right now.

Anyway, I think changing the eth0 LAN port back to WAN should be a separate commit, mentioning explicitly that it reverts the referenced one, and should obviously then do the same for er-x and er-x-sfp, as they have the same vendor OS/setup anyway.

pmelange replied:
Luckily I have an extra erx-sfp which is still running OEM firmware.  I just did a reset and the results are as you mentioned.  eth0 is statically assigned.  But all of the other ports are unreachable/unconfigured.  

So if OpenWRT really wants to stay with the default state of the erx and erx-sfp, then eth0 should be on LAN and all of the other ports should be disconnected and unreachable.  Putting everything on LAN is also, in this case, not the expected behaviour or "wrong".

I am willing to prepare a separate PR with the changes either to make eth0 the WAN port with eth1-X on LAN (preferred) or to have eth0 the only port on LAN with no WAN.

vDorst replied:
I have no option about this. I can keep it as it was or something else.

adrianschmutzler replied:
Since what you do in this PR is just adding another port, IMO this should be just changed to
`ucidef_set_interface_lan "eth0 eth1 eth2 eth3 eth4 eth5"`

pmelange replied:
I would say to just make ```ucidef_set_interface_lan "eth0 eth1 eth2 eth3 eth4 eth5"``` and I will submit a PR to create WAN on eth0.   Choosing which ports belong to which interface is not really related to enabling the SFP port.

vDorst replied:
@adrianschmutzler Fixed.

pmelange replied:
A new PR requesting a change to ucidef_set_interface_lan_wan can be found here.
#2961

pmelange commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -1141,3 +1141,23 @@ define KernelPackage/mlx5-core/description
+define KernelPackage/sfp
+  SUBMENU:=$(NETWORK_DEVICES_MENU)
+  TITLE:=SFP cage support
+  DEPENDS:=+kmod-i2c-core
+  KCONFIG:= \
+    CONFIG_SFP=m \
+    CONFIG_MDIO_I2C=m
```
Should this also depend on +kmod-hwmon-core and some form of KCONFIG too?

vDorst replied:
I added kmod-hwmon-core to the DEPENDS line.

vDorst commented:
Changes:
* rebased
* renamed `ephy5` to `ephy7`
* add `kmod-hwmon-core` to `kmod-sfp` dependency list.

vDorst commented:
Changed:
* Update PR message and title
* Define all interfaces as lan interfaces.

pmelange commented:
I have compiled and booted this image.  While the system was running, I inserted an SFP module.

```
[ 8920.454570] sfp sfp_eth5: module D-LINK           DEM-311GT        rev 1.0  sn SA0M1HC000017    dc 171114  
[ 8920.474003] Atheros 8031 ethernet mdio-bus:07: SFP interface 1000base-x
```

Then I turned on a switch with an SFP module a few seconds later, catching the link activity...

```
[ 9000.566582] mt7530 mdio-bus:1f eth5: Link is Down
[ 9001.598585] mt7530 mdio-bus:1f eth5: Link is Up - 1Gbps/Full - flow control off
[ 9001.613210] br-lan: port 2(eth5) entered blocking state
[ 9001.623653] br-lan: port 2(eth5) entered forwarding state
```
The network connection worked, I was able to ssh into the router via switch->SFP-port.

and later removed the module again
```
[ 9209.714600] sfp sfp_eth5: module removed
```

I hope this test will help

Thanks @vDorst for adding this support.

dengqf6 commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -1141,3 +1141,22 @@ define KernelPackage/mlx5-core/description
+  TITLE:=SFP cage support
+  DEPENDS:=+kmod-i2c-core +kmod-hwmon-core
+  KCONFIG:= CONFIG_SFP \
+	CONFIG_MDIO_I2C=y
```
Why is this `=y`?

vDorst replied:
I am not sure, when do I need to add `=m`, `=y`, `=n` or just the name.
If I look in the file `netsupport.mk` I see many combination.

Making it a module, this feels like the right line. But maybe I am wrong.
```
  KCONFIG:= \
        CONFIG_SFP=m \
        CONFIG_MDIO_I2C=m
```

dengqf6 replied:
`=y` means built-in
You can either use `=m` or just the name

vDorst replied:
Fixed both with `=m`.

vDorst commented:
Changes:
* Rebased
* kmod-sfp: set both kconfig options to `=m`.

lynxis commented on package/kernel/linux/modules/netdevices.mk:
```diff
@@ -1141,3 +1141,23 @@ define KernelPackage/mlx5-core/description
+  TITLE:=SFP cage support
+  DEPENDS:=+kmod-i2c-core +kmod-hwmon-core
+  KCONFIG:= \
+  	CONFIG_SFP=m \
```
can you replace these spaces with a tab?

vDorst replied:
Well spotted, Fixed
Thanks!

lynxis commented:
@vDorst looks good. Only a small nitpick, can you replace the spaces with a tab? I've tested it, works also with GBIC ethernet SFP. Thanks!

vDorst commented:
Changes:
* Rebased
* Removed spaces at begin of the line in `netdevices.mk`. Noticed by @lynxis 

FYI: There is a discussion going about DSA port names. http://lists.infradead.org/pipermail/openwrt-devel/2020-April/022988.html

lynxis commented:
`        Link partner advertised link modes:  1000baseX/Full `

lynxis commented:
merged. thanks!

john-tho commented:
Hi @vDorst @dengqf6, thank you very much for the mt7621 SFP support.

@bcanterac [has an SFP module](https://github.com/openwrt/openwrt/pull/3103#issuecomment-723730994) that runs at: [`sfp_quirk_2500basex`](https://github.com/torvalds/linux/commit/b0eae33b2583dceb36224619f9fd85e6140ae594). It worked for the (linux 3.3.5 based) OEM firmware, but is not functioning in OpenWrt, in a Mikrotik rb760igs.


```
[ 120.249527] sfp sfp1: module ALCATELLUCENT 3FE46541AA
[ 120.258421] sfp sfp1: Unknown/unsupported extended compliance code: 0x20
[ 120.265158] Atheros 8031 ethernet mdio-bus:07: incompatible SFP module inserted;Only SGMII/1000BASEX are supported!
```
[`Unknown/unsupported` is only a warning](https://elixir.bootlin.com/linux/v5.4.50/source/drivers/net/phy/sfp-bus.c#L209) here.

Is it possible to fall-through, or fallback 2500basex to 1000basex in [`ramips/patches-5.4/991-at803x.patch`](https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/patches-5.4/991-at803x.patch), or would it be better to set the quirk to 1000basex?

Cheers

vDorst commented:
@john-tho So the module really runs at `1000basex` but only reports `2500basex`?
If that is the case I should say we need to fix reporting on the SFP module side with maybe an other quirk.
So that the SFP modules reports the right modes that it supports and let kernels phylink api handles the rest.

What does `ethertool -m <sfp_interface>` and `ethertool <sfp_interface>` report?
Remove the serialnumbers if you don't want to report that.

Based on this repo https://github.com/minhng99/alcatel_lucent-lantiq_falcon.
The SFP module contains a microcontroller. So the at803x doesn't interface with the GPON interface directly but thru that microcontroller, which can run the interface between SPF module and AT803x at `1000basex` and `2500basex`.

bcanterac commented:
Hi, thanks for the blazing fast comments! To start with, I could not install ethertool on the RB760iGS so I installed ethtool which should be the same.

About the outputs, "ethtool -m sfp" gave me a hex dump, I guess this was not the idea right? Apart from that, here it is "ethtool sfp" output from the router:

``root@OpenWrt:~# ethtool sfp
Settings for sfp:
        Supported ports: [ TP MII ]
        Supported link modes:   100baseT/Half 100baseT/Full 
                                1000baseT/Full 
                                1000baseX/Full 
        Supported pause frame use: Symmetric Receive-only
        Supports auto-negotiation: Yes
        Supported FEC modes: Not reported
        Advertised link modes:  100baseT/Half 100baseT/Full 
                                1000baseT/Full 
                                1000baseX/Full 
        Advertised pause frame use: Symmetric Receive-only
        Advertised auto-negotiation: Yes
        Advertised FEC modes: Not reported
        Link partner advertised link modes:  1000baseX/Full 
        Link partner advertised pause frame use: Symmetric Receive-only
        Link partner advertised auto-negotiation: Yes
        Link partner advertised FEC modes: Not reported
        Speed: 1000Mb/s
        Duplex: Full
        Port: MII
        PHYAD: 7
        Transceiver: external
        Auto-negotiation: on
        Current message level: 0x000000ff (255)
                               drv probe link timer ifdown ifup rx_err tx_err
        Link detected: yes
``

Some comments about the SFP that may or may not be useful, it runs openWRT as well and I can access via SSH to it, I tried to make the same ethtool command on the SFP but it is not installed and the SFP itself has no internet. Also, the SFP requires to be connected to the optical fiber in order to be reachable to it's IP address, and apart from that it does not get "05" (GPON online status) despite being with the exact configuration as with OEM's router firmware. This should be independent of router firmware right?

If you need any extra info just tell me, I'm open to work with it to have the SFP up and running.

P.D.: @john-tho do you know why the RB760iGS image does not include luci? I had to install it manually and it runs fine, I don't know why it does not come preinstalled on the image.

john-tho commented:
> So the module really runs at 1000basex but only reports 2500basex?

@vDorst It looks like it can do both? One example shows module can autodetect LAN mode.

Before the kernel SFP quirk was implemented, I think kernel would have used `1000basex` for this module, based on the _transceiver_ `e1000_base_lx` module EEPROM value?

The quick and dirty functionality test for @bcanterac is to compile Openwrt with the [SFP quirk changed from](https://elixir.bootlin.com/linux/v5.4.75/source/drivers/net/phy/sfp-bus.c#L43) 2500 to 1000. @bcanterac, mention me _johnth_ on the openwrt forum if you have compile problems with this and I will help there.

https://elixir.bootlin.com/linux/v5.4.50/source/drivers/net/phy/sfp-bus.c#L255
`sfp_select_interface` will only return the fastest supported mode, so there seems no point having the quirk set both?

I do not have any SFP+ modules, to see what they report, and how they fall back.

I do not know if 2500BASEX is supported by the kernel mediatek soc driver?
It looks like it falls-through?
https://elixir.bootlin.com/linux/v5.4.66/source/drivers/net/ethernet/mediatek/mtk_eth_soc.c#L247
https://elixir.bootlin.com/linux/v5.4.66/source/drivers/net/ethernet/mediatek/mtk_sgmii.c#L72
```
		case PHY_INTERFACE_MODE_1000BASEX:
		case PHY_INTERFACE_MODE_2500BASEX:
		case PHY_INTERFACE_MODE_SGMII:
			if (MTK_HAS_CAPS(eth->soc->caps, MTK_SGMII)) {
				err = mtk_gmac_sgmii_path_setup(eth, mac->id);
				if (err)
					goto init_err;
			}
			break;
```


https://elixir.bootlin.com/linux/v5.4.75/source/drivers/net/phy/sfp-bus.c#L189
```
void sfp_parse_support

	br_nom = 3200;
	br_min = 0;
	br_max = 0;

	if (id->base.e1000_base_sx ||
	    id->base.e1000_base_lx ||
	    id->base.e1000_base_cx)
		phylink_set(modes, 1000baseX_Full);

	switch (id->base.extended_cc)
	default:
		dev_warn(bus->sfp_dev,
			 "Unknown/unsupported extended compliance code: 0x%02x\n",
			 id->base.extended_cc);
		break;
	}

	if (bus->sfp_quirk)
		bus->sfp_quirk->modes(id, modes);

	sfp_quirk_2500basex
		phylink_set(modes, 2500baseX_Full);
```

A good amount of data for this module here, with on-module, and Mikrotik-old-linux or Debian host information: https://forum.openwrt.org/t/will-gpon-nokia-g-010s-a-change-sn/69602/5

```
ONTUSER@SFP:~# onu lanpsg 0
errorcode=0 pport=0 mode=15 enable=1 link_status=4 phy_duplex=1

- enum lan_mode_interface mode
   LAN_MODE_OFF = 0
   LAN_MODE_GPHY = 1
   LAN_MODE_EPHY = 2
   LAN_MODE_SGMII = 3
   LAN_MODE_SGMII_FAST = 4
   LAN_MODE_RGMII_MAC = 5
   LAN_MODE_RMII_MAC = 6
   LAN_MODE_RMII_PHY = 7
   LAN_MODE_GMII_MAC = 8
   LAN_MODE_GMII_PHY = 9
   LAN_MODE_MII_MAC = 10
   LAN_MODE_MII_PHY = 11
   LAN_MODE_TMII_MAC = 12
   LAN_MODE_TMII_PHY = 13
   LAN_MODE_TBI_SERDES = 14
   LAN_MODE_TBI_AUTODETECT = 15
```

Example EEPROM dump from under pfSense / freeBSD:
https://forum.netgate.com/post/825851
```
00000000  03 04 01 00 00 00 02 00  00 00 00 03 20 00 28 ff  |............ .(.|
00000010  00 00 00 00 41 4c 43 41  54 45 4c 4c 55 43 45 4e  |....ALCATELLUCEN|
00000020  54 20 20 20 20 20 20 20  33 46 45 34 36 35 34 31  |T       3FE46541|
00000030  41 41 20 20 20 20 20 20  30 30 30 31 05 1e ff dc  |AA      0001....|
00000040  00 1a 00 00 41 4c 43 4c  46 38 38 35 31 39 45 34  |....ALCLF88519E4|
00000050  20 20 20 20 31 38 30 38  32 39 20 20 68 f0 05 5d  |    180829  h..]|
00000060  41 4c 43 41 54 45 4c 20  33 46 45 34 36 35 34 31  |ALCATEL 3FE46541|
00000070  41 41 30 31 32 42 56 4c  33 41 38 4a 4e 41 41 97  |AA012BVL3A8JNAA.|
00000080
```
From https://git.kernel.org/pub/scm/network/ethtool/ethtool.git/tree/sfpid.c, this gives:

|Field|Value|
|------|--------|
|ID|SFF8024_ID_SFP|
|eID|GBIC/SFP defined by 2-wire interface ID|
|connector|SFF8024_CTOR_SC|
|transceiver|Ethernet: 1000BASE-LX `0x6=0x2`|
|encoding|SFF8024_ENCODING_NRZ|
|BR nom.|3200MBd `0xc=0x20 * 100`|
|rate|unspecified|
|Length (SMF,km)|40|
|Length (SMF,m)|25500|
|Length (50um,m)|0|
|Length (62.5um,m)|0|
|Length (Copper,m)|0|
|Length (OM3,m)|0|
|Vendor|ALCATELLUCENT|
|escc `0x24`|Extended: 100G SWDM4 `0x20` (nonsense)|
|Vendor OUI||
|Vendor part|3FE46541AA|
|Vendor rev.|0001|
|Laser wavelength (nm) `0x3c<<8\|0x3d`|1310 `(0x5<<8)\|0x1e`|
|Options `0x40, 0x41`|Has RX_LOS, TX_FAULT, TX_DISABLE `0x41=0x1a == 1<<4 \| 1<<3 \| 1<<1`|
|BR margin max. `0x42`|0|
|BR margin min. `0x43`|0|
|Vendor serial `0x44…0x53`|ALCLF88519E4|
|Date `0x54…0x5b`|180829|


___

> `ethtool -m sfp` gave me a hex dump

Could you please post the hex dump for us if anything apart from the date and serial are different to the above for your module? You can x out your serial if you want to.

>  image does not include luci

@bcanterac No `uhttpd` / luci in Openwrt snapshots. You can use backticks: ``` around terminal output / code to leave it unformatted: 3 for multiline, or one pair around inline code.

Cheers

bcanterac commented:
@john-tho it would be great if you can compile openwrt with this modification, so I can install it and test exactly the sfp quirk change, I am not so sure how to compile it in the exact way that there is needed and don't want to make wrong tests.

Meanwhile I will reconnect everything and take the hex dump.

Cheers!

vDorst commented:
> > So the module really runs at 1000basex but only reports 2500basex?
> 
> @vDorst It looks like it can do both? One example shows module can autodetect LAN mode.

I think that is true because the microcontroller inside the SFP can handles both.

> Before the kernel SFP quirk was implemented, I think kernel would have used `1000basex` for this module, based on the _transceiver_ `e1000_base_lx` module EEPROM value?
> 
> The quick and dirty functionality test for @bcanterac is to compile Openwrt with the [SFP quirk changed from](https://elixir.bootlin.com/linux/v5.4.75/source/drivers/net/phy/sfp-bus.c#L43) 2500 to 1000. @bcanterac, mention me _johnth_ on the openwrt forum if you have compile problems with this and I will help there.


I am not sure but phylink should automaticly select the the right modes that both side support.
Adding `2500baseX_Full` to the mtk_eth_soc or the at803x driver seems not the right path to go.

I see multiple thing you can try:

1. Maybe the mtk_soc_eth is wrong by this phylink_helper function https://elixir.bootlin.com/linux/v5.4.66/source/drivers/net/ethernet/mediatek/mtk_eth_soc.c#L543
   Just comment that out for testing.

2. Modify the sfp_quirk_2500basex so that it also reports `1000baseX_Full` in https://elixir.bootlin.com/linux/v5.4.75/source/drivers/net/phy/sfp-bus.c#L43.
	Like this for the test.

	```
	static void sfp_quirk_2500basex(const struct sfp_eeprom_id *id,
					unsigned long *modes)
	{
		phylink_set(modes, 1000baseX_Full);
		phylink_set(modes, 2500baseX_Full);
	}
	```

> 
> https://elixir.bootlin.com/linux/v5.4.50/source/drivers/net/phy/sfp-bus.c#L255
> `sfp_select_interface` will only return the fastest supported mode, so there seems no point having the quirk set both?
> 
> I do not have any SFP+ modules, to see what they report, and how they fall back.
> 
> I do not know if 2500BASEX is supported by the kernel mediatek soc driver?
> It looks like it falls-through?
> https://elixir.bootlin.com/linux/v5.4.66/source/drivers/net/ethernet/mediatek/mtk_eth_soc.c#L247
> https://elixir.bootlin.com/linux/v5.4.66/source/drivers/net/ethernet/mediatek/mtk_sgmii.c#L72
> 
> ```
> 		case PHY_INTERFACE_MODE_1000BASEX:
> 		case PHY_INTERFACE_MODE_2500BASEX:
> 		case PHY_INTERFACE_MODE_SGMII:
> 			if (MTK_HAS_CAPS(eth->soc->caps, MTK_SGMII)) {
> 				err = mtk_gmac_sgmii_path_setup(eth, mac->id);
> 				if (err)
> 					goto init_err;
> 			}
> 			break;
> ```
> 

Only the MT7622 and MT7629 ARM Soc has a SGMII bus.

On the Ubiquiti ER-X-SFP and Mikrotik diagram looks like this:

`MT7621 Soc -> GMAC RGMII2 -> RGMII BUS -> AT8033 CHIP; Act as RGMII to 1000BASEX converter -> 1000BASEX -> SFP Cage`

I think it get even weirder with the GPON module.

`MT7621 Soc -> GMAC RGMII2 -> RGMII BUS -> AT8033 CHIP; Act as RGMII to 1000BASEX converter -> 1000BASEX or 2500BASEX -> GPON Module -> 1000BASEX or 2500BASEX -> GPON PHY LASER`.

I hope I have some time soon to test it with my setup.
I don't own any of these modules but I think I can also make a fake quirk for my module to test it out.

john-tho commented:
> make a fake quirk for my module to test it out.

Good point. I added a 2500baseX_Full quirk to one of my SFP modules.
I needed to comment out the `991-at803x.patch sfp_select_interface (SGMII | 1000BASEX)` test `return` for it to function. https://github.com/openwrt/openwrt/blob/a47279154e08d54df05fa8bf45fe935ebf0df5da/target/linux/ramips/patches-5.4/991-at803x.patch#L73

> I am not sure but phylink should automaticly select the the right modes that both side support.
> Adding `2500baseX_Full` to the mtk_eth_soc or the at803x driver seems not the right path to go.

Agreed

As is, SFP module load is aborted by `sfp_select_interface` in our at803x modification, as 2500base-X will always be returned first (highest first) if supported by the module: https://elixir.bootlin.com/linux/v5.4.75/source/drivers/net/phy/sfp-bus.c#L342 https://github.com/openwrt/openwrt/blob/a47279154e08d54df05fa8bf45fe935ebf0df5da/target/linux/ramips/patches-5.4/991-at803x.patch#L67



If we _need_ to test SFP module compatibility here, we could directly test `sfp_parse_support support` against our acceptable link modes? There is probably a better way to do this; the identifiers are used by a few drivers: https://elixir.bootlin.com/linux/v5.4.75/A/ident/ETHTOOL_LINK_MODE_1000baseT_Full_BIT
```
--- linux-5.4.74/drivers/net/phy/at803x.c.orig	2020-11-11 08:28:12.572722419 +1000
+++ linux-5.4.74/drivers/net/phy/at803x.c	2020-11-11 11:17:17.804118116 +1000
@@ -269,20 +269,22 @@
 {
 	struct phy_device *phydev = upstream;
 	__ETHTOOL_DECLARE_LINK_MODE_MASK(support) = { 0, };
-	phy_interface_t iface;
 
 	sfp_parse_support(phydev->sfp_bus, id, support);
-	iface = sfp_select_interface(phydev->sfp_bus, support);
 
-	if (iface != PHY_INTERFACE_MODE_SGMII &&
-	    iface != PHY_INTERFACE_MODE_1000BASEX) {
+	dev_info(&phydev->mdio.dev, "SFP module supported link mode bits "
+		 "(ethtool_link_mode_bit_indices): "
+                 "*pbl %*pbl\n", __ETHTOOL_LINK_MODE_MASK_NBITS, support);
+
+	if (linkmode_test_bit(ETHTOOL_LINK_MODE_1000baseT_Half_BIT, support) ||
+	    linkmode_test_bit(ETHTOOL_LINK_MODE_1000baseT_Full_BIT, support) ||
+	    linkmode_test_bit(ETHTOOL_LINK_MODE_1000baseX_Full_BIT, support)) {
+	} else {
 		dev_info(&phydev->mdio.dev, "incompatible SFP module inserted;"
 			 "Only SGMII/1000BASEX are supported!\n");
 		return -EINVAL;
 	}
 
-	dev_info(&phydev->mdio.dev, "SFP interface %s", phy_modes(iface));
-
 	return 0;
 }

``` 

Thank you for the hardware flow diagrams!

My fibre LC SFP module, quirked to use 2500base-X
Works fine with either:
current ar803x SFP module interface test return disabled:
```
[    9.859189] sfp sfp1: module FS               SFP1G-LX-31
[    9.881767] SFP module quirked to 2500base-X
[    9.899324] Atheros 8031 ethernet mdio-bus:07: incompatible SFP module inserted;Only SGMII/1000BASEX are supported! Got: 19
[    9.910414] Atheros 8031 ethernet mdio-bus:07: SFP interface 2500base-x
```
My changes ar803x SFP
```
[    9.860656] sfp sfp1: module FS               SFP1G-LX-31
[    9.870045] SFP module quirked to 2500base-X
[    9.870062] Atheros 8031 ethernet mdio-bus:07: SFP module supported link mode bits (ethtool_link_mode_bit_indices): *pbl 6,13-15,41
```
Still works with `phylink_zero(modes);` before the quirk is applied, and `at803x_sfp_insert return 0` to force `2500base-X` only in SFP linkmode. Guess next step would be to quirk to something nonsense to test if the check is needed in `at803x_sfp_insert`?
```
[    9.854386] sfp sfp1: module FS               SFP1G-LX-31
[    9.863766] SFP module quirked to 2500base-X
[    9.863783] Atheros 8031 ethernet mdio-bus:07: SFP module supported link mode bits (ethtool_link_mode_bit_indices): *pbl 6,13-15
[    9.879563] Atheros 8031 ethernet mdio-bus:07: incompatible SFP module inserted;Only SGMII/1000BASEX are supported!
```
___

>        ```
>        static void sfp_quirk_2500basex(const struct sfp_eeprom_id *id,
>        				unsigned long *modes)
>        {
>        	phylink_set(modes, 1000baseX_Full);
>        	phylink_set(modes, 2500baseX_Full);
>        }
>        ```

I tested before I saw this with @bcanterac, and the GPON worked with the quirk changed from 2500 to 1000.
I _think_ [`1000baseX_Full` should be set by the kernel based on module's reported transceiver `e1000_base_lx`](https://elixir.bootlin.com/linux/v5.4.75/source/drivers/net/phy/sfp-bus.c#L226). This is what happens with my test-quirked module.

Let me know what you think?

Cheers

vDorst commented:
@john-tho  I also played with it.

I think this is the right way.
```
diff --git a/drivers/net/phy/at803x.c b/drivers/net/phy/at803x.c
index 8fbb78ec2ab46..cadaf6cd9ee90 100644
--- a/drivers/net/phy/at803x.c
+++ b/drivers/net/phy/at803x.c
@@ -20,6 +20,7 @@
 #include <linux/regulator/consumer.h>
 #include <dt-bindings/net/qca-ar803x.h>
 #include <linux/sfp.h>
 #include <linux/regulator/consumer.h>
 #include <dt-bindings/net/qca-ar803x.h>
 #include <linux/sfp.h>
+#include <linux/phylink.h>

 #define AT803X_SPECIFIC_STATUS                 0x11
 #define AT803X_SS_SPEED_MASK                   (3 << 14)
@@ -500,23 +501,41 @@ static int at803x_mode(struct phy_device *phydev)

 static int at803x_sfp_insert(void *upstream, const struct sfp_eeprom_id *id)
 {
-       struct phy_device *phydev = upstream;
+       __ETHTOOL_DECLARE_LINK_MODE_MASK(at803x_support) = { 0, };
        __ETHTOOL_DECLARE_LINK_MODE_MASK(support) = { 0, };
+       struct phy_device *phydev = upstream;
        phy_interface_t iface;

+       /* AT803x only support 1000baseX but SGMII works fine when module runs
+        * at 1Gbit.
+        */
+       phylink_set(at803x_support, 1000baseT_Full);
+       phylink_set(at803x_support, 1000baseX_Full);
+
        sfp_parse_support(phydev->sfp_bus, id, support);
+
+       // Filter out interface that both sides supports
+       linkmode_and(support, support, at803x_support);
+
+       if (linkmode_empty(support)) {
+           goto unsupported_mode;
+       }
+
        iface = sfp_select_interface(phydev->sfp_bus, support);

        if (iface != PHY_INTERFACE_MODE_SGMII &&
            iface != PHY_INTERFACE_MODE_1000BASEX) {
-               dev_info(&phydev->mdio.dev, "incompatible SFP module inserted;"
-                        "Only SGMII/1000BASEX are supported!\n");
-               return -EINVAL;
+           goto unsupported_mode;
        }

        dev_info(&phydev->mdio.dev, "SFP interface %s", phy_modes(iface));

        return 0;
+
+unsupported_mode:
+       dev_info(&phydev->mdio.dev, "incompatible SFP module inserted;"
+                "Only SGMII at 1Gbit/1000BASEX are supported!\n");
+       return -EINVAL;
 }
```

john-tho commented:
> I think this is the right way.
> Get SPF modules supported interfaces and logic AND it with the interfaces that the at803x supports.
> Check if we have one or more that both sides supports, if so ask `sfp_parse_support()` for a mode.
> Also check that and return or report an error.

I like it. Looks good to me.
To be pedantic, I don't like _Filter out_ in the `linkmode_and` comment:
```
-       // Filter out interface that both sides supports
+       // Limit to interfaces that both sides support
```
Will you create the Openwrt quilt patch, and open a PR, or mail it in? Otherwise, let me know if you are short on time and I can?

> I looked at Russell King's his phylink patches. http://git.armlinux.org.uk/cgit/linux-arm.git/log/?h=net-queue
> I see that he is working on this issue.

Good find!

Have you considered sending your 991 patch upstream?

Thanks

vDorst commented:
I shall change the comment.

I make a PR tomorrow.

> Have you considered sending your 991 patch upstream?

Yes, but duo issue I have with the upstream kernel, it lost my interested.
Issue was that the upstream kernel locked-up at boot at random places without an backtrace.

vDorst commented:
@john-tho @bcanterac 
Created PR: https://github.com/openwrt/openwrt/pull/3584
